File: Java/src/com/yourtion/Pattern09/CalculatorExample.java
package com.yourtion.Pattern09;

/**
 * Created by Yourtion on 28/11/2016.
 */
public class CalculatorExample {

    public void run() {
        System.out.println("Without Logging :");

        CalculatorImpl cal1 = new CalculatorImpl();
        System.out.println(cal1.Add(1, 2));
        System.out.println(cal1.Subtract(4,3));
        System.out.println(cal1.Divide(6,2));
        System.out.println(cal1.Multiply(2,3));

        System.out.println("");
        System.out.println("Logging :");

        CalculatorImpl cal2 = new LoggingCalculator();
        System.out.println(cal2.Add(1,2));
        System.out.println(cal2.Subtract(4,3));
        System.out.println(cal2.Divide(6,2));
        System.out.println(cal2.Multiply(2,3));

    }
}

File: Scala/src/com/yourtion/Pattern09/CalculatorExample.scala
package com.yourtion.Pattern09

/**
  * Created by Yourtion on 28/11/2016.
  */
object CalculatorExample {

    def add(a: Int, b: Int) = a + b
    def subtract(a: Int, b: Int) = a - b
    def multiply(a: Int, b: Int) = a * b
    def divide(a: Int, b: Int) = a / b

    def makeLogger(calcFn: (Int, Int) => Int) =
      (a: Int, b: Int) => {
        val result = calcFn(a, b)
        println("Result is: " + result)
        result
      }

    val loggingAdd = makeLogger(add)
    val loggingSubtract = makeLogger(subtract)
    val loggingMultiply = makeLogger(multiply)
    val loggingDivide = makeLogger(divide)

  def run(): Unit = {
    println("Without Logging :")

    println(add(1,2))
    println(subtract(4,3))
    println(divide(6,2))
    println(multiply(2,3))

    println()
    println("Logging :")
    println(loggingAdd(1,2))
    println(loggingSubtract(4,3))
    println(loggingDivide(6,2))
    println(loggingMultiply(2,3))

  }

}
